hoomd/md/GeneralEnvelope.h

```diff
@@ -0,0 +1,310 @@
+#define HOSTDEVICE __host__ __device__
+#else
+#define DEVICE
+#define HOSTDEVICE
+#endif
+
+namespace hoomd
+    {
+namespace md
+    {
+
+/*
+  The GeneralEnvelope creates the pair potential modulator.
+*/
+
+        static inline std::string vecString(vec3<Scalar> a) {
+            return std::to_string(a.x) + ", " + std::to_string(a.y) + ", " + std::to_string(a.z) + '\n';
+        }
+
+
+class GeneralEnvelope
+{
+public:
+    struct param_type
+    {
+        param_type()
+            {
+            }
+#ifndef __HIPCC__
+        param_type(pybind11::dict params) // param dict can take any python type
+            {
+                cosalpha = fast::cos(params["alpha"].cast<Scalar>());
+                omega = params["omega"].cast<Scalar>();
+            }
+
+        pybind11::dict toPython()
+            {
+                pybind11::dict v;
+                v["alpha"] = fast::acos(cosalpha);
+                v["omega"] = omega;
+                return v;
+            }
+#endif
+        Scalar cosalpha;
+        Scalar omega;
+    }__attribute__((aligned(16)));
+
+    struct shape_type
+    {
+        HOSTDEVICE shape_type() { }
+
+#ifndef __HIPCC__
+
+        shape_type(pybind11::object patch_location)
+            {
+            pybind11::tuple p_py = patch_location;
+            if (len(p_py) != 3)
+                throw std::runtime_error("Each patch position must have 3 elements");
+            vec3<Scalar> p = vec3<Scalar>(pybind11::cast<Scalar>(p_py[0]),
+                                          pybind11::cast<Scalar>(p_py[1]),
+                                          pybind11::cast<Scalar>(p_py[2]));
+
+            // normalize
+            p = p * fast::rsqrt(dot(p, p));
+            m_norm_patch_local_dir = vec_to_scalar3(p);
+            }
+
+        pybind11::object toPython()
+            {
+            return pybind11::make_tuple(m_norm_patch_local_dir.x, m_norm_patch_local_dir.y, m_norm_patch_local_dir.z);
+            }
+#endif
+
+        Scalar3 m_norm_patch_local_dir;
+    };
+
+    DEVICE GeneralEnvelope( // TODO replace GeneralEnvelope with PatchesEnvelope
+        const Scalar3& _dr,
+        const Scalar4& _quat_i, // Note in hoomd, the quaternion is how to get from the particle orientation to align to the world orientation. so World = qi Local qi-1
+        const Scalar4& _quat_j,
+        const Scalar _rcutsq,
+        const param_type& _params,
+        const shape_type& shape_i,
+        const shape_type& shape_j)
+// #ifdef __HIPCC__
+        // TODO decide which is faster on GPU and CPU. // use 250_000 particles to saturate GPU // benchmark on brett
+        // : dr(_dr), qi(_quat_i), qj(_quat_j), params(_params), p_i(shape_i.m_norm_patch_local_dir), p_j(shape_j.m_norm_patch_local_dir)
+// #else
+        : dr(_dr), R_i(rotmat3<ShortReal>((quat<ShortReal>)_quat_i)), R_j(rotmat3<ShortReal>((quat<ShortReal>)_quat_j)), params(_params), p_i(shape_i.m_norm_patch_local_dir), p_j(shape_j.m_norm_patch_local_dir)
```

Comment:
Unless you have benchmarks that show otherwise, I think all the `ShortReals` should be `LongReal`. In most cases, converting back and forth between `float` and `double` is far more expensive than the cost savings of a few math operations in `float`.